File: algorithm/leetcode/classical150item/hash/longestConsecutive.go
package main

import "go-career/algorithm/leetcode"

type exampleLongestConsecutive struct {
	nums   []int
	answer int
}

func main() {
	s := []exampleLongestConsecutive{
		{leetcode.ToIntSlice("[100,4,200,1,3,2]"), 4},
	}
	for _, v := range s {
		leetcode.Printf(longestConsecutive(v.nums), v.answer)
	}
}
func longestConsecutive(nums []int) int {
	m := make(map[int]struct{}, len(nums))
	for _, v := range nums {
		m[v] = struct{}{}
	}
	longest := 0
	for k := range m {
		long := 0
		step, forward := 1, true
	loop:
		for _, ok := m[k]; ok; _, ok = m[k] {
			long++
			delete(m, k)
			k += step
		}
		if forward {
			forward = false
			k, step = k-long-1, -1
			goto loop
		}
		if long > longest {
			longest = long
		}
	}
	return longest
}

File: algorithm/leetcode/classical150item/hash/longestConsecutive.py
from typing import List


class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        d = dict.fromkeys(nums, True)
        longest = 0
        for number in d.keys():
            if number - 1 in d:
                continue
            long = 0
            while number in d.keys():
                number += 1
                long += 1
            else:
                longest = long if long > longest else longest
        return longest


if __name__ == "__main__":
    examples = [
        ([9, 1, 4, 7, 3, -1, 0, 5, 8, -1, 6], 7),
        ([100, 4, 200, 1, 3, 2], 4),
    ]
    solution = Solution()
    for nums, answer in examples:
        print(f"{solution.longestConsecutive(nums)}\n{answer}\n")
